File: ts/20.valid-parentheses.ts
/*
 * @lc app=leetcode id=20 lang=typescript
 *
 * [20] Valid Parentheses
 */

// @lc code=start
function isValid(s: string): boolean {
  const stack = [];

  for (let i = 0; i < s.length; i++) {
    if (OPEN_SET.has(s[i])) {
      stack.push(s[i]);
    } else if (PARENTHESES_MAP.get(s[i]) === stack[stack.length - 1]) {
      stack.pop();
    } else {
      return false;
    }
  }

  return stack.length === 0;
}

const OPEN_SET = new Set(["(", "{", "["]);

const PARENTHESES_MAP = new Map([
  [")", "("],
  ["}", "{"],
  ["]", "["],
]);
// @lc code=end

File: go/20.valid-parentheses.go
/*
 * @lc app=leetcode id=20 lang=golang
 *
 * [20] Valid Parentheses
 */
package leetcode

// @lc code=start
func isValid(s string) bool {
	pMap := map[rune]rune{
		')': '(',
		']': '[',
		'}': '{',
	}

	stack := make([]rune, 0, 1)

	for _, char := range s {
		switch char {
		case '(', '[', '{':
			stack = append(stack, char)
		case ')', ']', '}':
			if len(stack) == 0 || stack[len(stack)-1] != pMap[char] {
				return false
			} else {
				stack = stack[:len(stack)-1]
			}
		}
	}

	return len(stack) == 0
}

// @lc code=end
